flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/AbstractFileSource.java

```diff
@@ -0,0 +1,321 @@
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The base class for File Sources. The main implementation to use is the {@link FileSource},
+ * which also has the majority of the documentation.
+ *
+ * <p>To read new formats, one commonly does NOT need to extend this class, but should implement
+ * a new Format Reader (like {@link StreamFormat}, {@link BulkFormat}, {@link FileRecordFormat})
+ * and use it with the {@code FileSource}.
+ *
+ * <p>The only reason to extend this class is when a source needs a different type of
+ * <i>split</i>, meaning an extension of the {@link FileSourceSplit} to carry additional
+ * information.
+ *
+ * @param <T> The type of the events/records produced by this source.
+ * @param <SplitT> The subclass type of the FileSourceSplit used by the source implementation.
+ */
+@PublicEvolving
+public abstract class AbstractFileSource<T, SplitT extends FileSourceSplit>
+		implements Source<T, SplitT, PendingSplitsCheckpoint<SplitT>>, ResultTypeQueryable<T> {
+
+	private static final long serialVersionUID = 1L;
+
+	private final Path[] inputPaths;
+
+	private final FileEnumerator.Provider enumeratorFactory;
+
+	private final FileSplitAssigner.Provider assignerFactory;
+
+	private final BulkFormat<T, SplitT> readerFormat;
+
+	@Nullable
+	private final ContinuousEnumerationSettings continuousEnumerationSettings;
+
+	// ------------------------------------------------------------------------
+
+	protected AbstractFileSource(
+			final Path[] inputPaths,
+			final FileEnumerator.Provider fileEnumerator,
+			final FileSplitAssigner.Provider splitAssigner,
+			final BulkFormat<T, SplitT> readerFormat,
+			@Nullable final ContinuousEnumerationSettings continuousEnumerationSettings) {
+
+		checkArgument(inputPaths.length > 0);
+		this.inputPaths = inputPaths;
+		this.enumeratorFactory = checkNotNull(fileEnumerator);
+		this.assignerFactory = checkNotNull(splitAssigner);
+		this.readerFormat = checkNotNull(readerFormat);
+		this.continuousEnumerationSettings = continuousEnumerationSettings;
+	}
+
+	// ------------------------------------------------------------------------
+	//  Source API Methods
+	// ------------------------------------------------------------------------
+
+	@Override
+	public Boundedness getBoundedness() {
+		return continuousEnumerationSettings == null ? Boundedness.BOUNDED : Boundedness.CONTINUOUS_UNBOUNDED;
+	}
+
+	@Override
+	public SourceReader<T, SplitT> createReader(SourceReaderContext readerContext) {
+		return new FileSourceReader<>(readerContext, readerFormat, readerContext.getConfiguration());
+	}
+
+	@Override
+	public SplitEnumerator<SplitT, PendingSplitsCheckpoint<SplitT>> createEnumerator(
```

Comment:
I can image sub class override here:
Iceberg: Override all, `return new IcebergEnumerator`.
Hive: 
```
if (continuousEnumerationSettings != null && isPartitionedTable) {
    return new HiveSplitEnumerator;
} else {
   return super.createEnumerator();
}
```

Reply:
Agree, that can work.